File: 140-word-break-ii/word-break-ii.cpp
class Solution {
public:
    void find(int i, string temp, string &s, set<string>&st, vector<string>&ans){
        if(i>=s.size()){
            temp.pop_back();
            ans.push_back(temp);
            return ;
        }
        for(int j=i; j<s.size(); j++){
            if(st.find(s.substr(i, j-i+1))!=st.end()){
                find(j+1, temp+s.substr(i, j-i+1)+" ", s, st, ans);
            }
        }
        
    }
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        vector<string>ans;
        set<string>st;
        for(auto i:wordDict){
            st.insert(i);
        }
        find(0, "", s, st, ans);
        return ans;
    }
};

File: 140-word-break-ii/word-break-ii.java
class Solution {
    List<String> res = new ArrayList<String>();
    HashSet<String> set;
    public List<String> wordBreak(String s, List<String> wordDict) {
       set = new HashSet<String>(wordDict);
       fun(s,0,"");
       return res;
    }
    void fun(String s, int idx,String temp){

        if(idx == s.length()){
            // if(res.size() != 0)return;
            String t = temp.substring(0,temp.length()-1);
            res.add(t);
            return;
        }

        for(int i = idx; i<s.length(); i++){
            if(!set.contains(s.substring(idx,i+1)))continue;
            // ÷z.add(s.substring(idx,i+1));
            fun(s,i+1,temp+s.substring(idx,i+1)+" ");
            // temp.remove(temp.size()-1);
        }
    }
}